README.md
# indi-web

A simple web based INDI control panel.

![INDI Control Panel](doc/screenshot.png "INDI Control Panel")

## Running

```bash
git clone https://github.com/rickbassham/indi-web.git
cd indi-web
docker build -t indi-web:latest .
docker run \
    -e INDI_HOST=host.docker.internal \
    -e INDI_PORT=7624 \
    -p 8081:8081 \
    -it indi-web:latest
```

Now browse to http://localhost:8081

## Docker

This is also published to Docker Hub.

https://hub.docker.com/r/rickbassham/indi-web

![Docker Image Version (tag latest semver)](https://img.shields.io/docker/v/rickbassham/indi-web/latest)


CHANGELOG.md
# [1.1.0](https://github.com/rickbassham/indi-web/compare/v1.0.4...v1.1.0) (2020-12-24)


### Features

* handle indi server connection state ([f11cac6](https://github.com/rickbassham/indi-web/commit/f11cac69324cf46c1495916d122640e7f4728edd))

## [1.0.4](https://github.com/rickbassham/indi-web/compare/v1.0.3...v1.0.4) (2020-12-24)


### Bug Fixes

* code cleanup ([306c298](https://github.com/rickbassham/indi-web/commit/306c298651d7b6463aacd4492c88ead1f9ff47d2))

## [1.0.3](https://github.com/rickbassham/indi-web/compare/v1.0.2...v1.0.3) (2020-12-23)


### Bug Fixes

* update readme and bump version to test docker caching ([45069d0](https://github.com/rickbassham/indi-web/commit/45069d09ea9f95ed2f36fa3bfef578fb8d0ba0a0))

## [1.0.2](https://github.com/rickbassham/indi-web/compare/v1.0.1...v1.0.2) (2020-12-23)


### Bug Fixes

* add support for sexagesimal numbers ([36270aa](https://github.com/rickbassham/indi-web/commit/36270aa81c666b7efe72bee85413aa81340e4f00))

## [1.0.1](https://github.com/rickbassham/indi-web/compare/v1.0.0...v1.0.1) (2020-12-23)


### Bug Fixes

* fix banner color ([9e45b6c](https://github.com/rickbassham/indi-web/commit/9e45b6c2699e6e5204fd8202e2a5710d6ba5afd0))

# 1.0.0 (2020-12-23)


### Bug Fixes

* fix releaserc ([888a4ee](https://github.com/rickbassham/indi-web/commit/888a4ee31464dccca8a18e321a3630318b58e62b))


### Features

* a simple web based INDI control panel ([34b7727](https://github.com/rickbassham/indi-web/commit/34b7727cdefed18287d524ea2d8cf0e38667c886))


web/src/plugins/vuex.js
import Vue from 'vue'
import Vuex from 'vuex'

import indiStore from '@/plugins/stores/indi'

Vue.use(Vuex);

export default new Vuex.Store({
  modules: {
    indi: indiStore
  },
  state: {
    socket: {
      connected: false,
      error: null,
      reconnectCount: 0,
    },
  },
  mutations: {
    SOCKET_ONOPEN(state, event) {
      Vue.prototype.$socket = event.currentTarget
      state.socket.isConnected = true
      this.dispatch("indi/subscribe");
    },
    SOCKET_ONCLOSE(state) {
      state.socket.isConnected = false
    },
    SOCKET_ONERROR(state, event) {
      state.socket.error = event;
    },
    // default handler called for all methods
    SOCKET_ONMESSAGE(state, message) {
      // turn this into a dumb object
      message = JSON.parse(JSON.stringify(message));
      state.socket.message = message

      this.commit(`indi/${message.type}`, message.payload);
    },
    // mutations for reconnect methods
    SOCKET_RECONNECT(state, count) {
      state.socket.reconnectCount = count;
      state.socket.isConnected = true
    },
    SOCKET_RECONNECT_ERROR(state) {
      state.socket.reconnectError = true;
    },
  },
});


web/src/plugins/stores/indi.js
import Vue from 'vue'

import {
  floor,
  merge,
} from 'lodash'

import printf from "printf"

import {
  enableBLOB
} from 'indi-client';

const formatSexaNumber = (value, w, fracbase) => {
  const isNeg = value < 0;
  if (isNeg) {
    value = -value;
  }

  const n = floor(value * fracbase + 0.5);
  const d = floor(n / fracbase);
  const f = n % fracbase;

  let out = "";

  if (isNeg && d == 0) {
    out += printf(`%${w - 2}s-0`, "");
  } else {
    out += printf(`%${w}d`, isNeg ? -d : d);
  }

  let m = 0;
  let s = 0;

  switch (fracbase) {
    case 60:
      m = floor(f / (fracbase / 60));
      out += printf(":%02d", m);
      break;
    case 600:
      out += printf(":%02d.%1d", floor(f / 10), f % 10);
      break;
    case 3600:
      m = floor(f / (fracbase / 60));
      s = f % (fracbase / 60);
      out += printf(":%02d:%02d", m, s);
      break;
    case 36000:
      m = floor(f / (fracbase / 60));
      s = f % (fracbase / 60);
      out += printf(":%02d:%02d.%1d", m, floor(s / 10), s % 10);
      break;
    case 360000:
      m = floor(f / (fracbase / 60));
      s = f % (fracbase / 60);
      out += printf(":%02d:%02d.%02d", m, floor(s / 100), s % 100);
      break;
    default:
      console.warn(`unknown fracbase ${fracbase}`);
      return value;
  }

  return out;
}

const formatNumberValue = (fmt, value) => {
  if (!fmt) {
    return value;
  }

  const checkResult = /%(\d+)\.(\d+)m/.exec(fmt);

  if (checkResult) {
    const w = parseInt(checkResult[1]);
    const f = parseInt(checkResult[2]);

    let s = 60;

    switch (f) {
      case 9:
        s = 360000;
        break;
      case 8:
        s = 36000;
        break;
      case 6:
        s = 3600;
        break;
      case 5:
        s = 600;
        break;
    }

    return formatSexaNumber(value, w - f, s);
  } else {
    return printf(fmt, value);
  }
};

const applyUpdate = (state, payload) => {
  payload = JSON.parse(JSON.stringify(payload));

  const newData = {};
  newData.name = payload.device;
  newData[payload.name] = payload;
  const existing = state.devices[payload.device] ? JSON.parse(JSON.stringify(state.devices[payload.device])) : {};
  const merged = merge(existing, newData);

  if (payload.switches) {
    merged[payload.name].propType = "switch";
  }
  if (payload.numbers) {
    merged[payload.name].propType = "number";
    merged[payload.name].numbers = merged[payload.name].numbers.map(n => {
      return {
        ...n,
        formatted: formatNumberValue(n.format, n.value)
      }
    })
  }
  if (payload.texts) {
    merged[payload.name].propType = "text";
  }
  if (payload.lights) {
    merged[payload.name].propType = "light";
  }
  if (payload.blobs) {
    merged[payload.name].propType = "blob";
  }

  Vue.set(state.devices, payload.device, merged);
}

export default {
  namespaced: true,
  state: () => ({
    devices: {},
    error: null,
    status: 'disconnected',
  }),
  mutations: {
    ["message"]() {
    },
    ["setNumberVector"](state, payload) {
      applyUpdate(state, payload);
    },
    ["setTextVector"](state, payload) {
      applyUpdate(state, payload);
    },
    ["setSwitchVector"](state, payload) {
      applyUpdate(state, payload);
    },
    ["setBLOBVector"](state, payload) {
      applyUpdate(state, payload);
    },
    ["setLightVector"](state, payload) {
      applyUpdate(state, payload);
    },
    ["defTextVector"](state, payload) {
      applyUpdate(state, payload);
    },
    ["defSwitchVector"](state, payload) {
      applyUpdate(state, payload);
    },
    ["defNumberVector"](state, payload) {
      applyUpdate(state, payload);
    },
    ["defLightVector"](state, payload) {
      applyUpdate(state, payload);
    },
    ["defBLOBVector"](state, payload) {
      applyUpdate(state, payload);
    },
    ["newTextVector"](state, payload) {
      applyUpdate(state, payload);
    },
    ["newSwitchVector"](state, payload) {
      applyUpdate(state, payload);
    },
    ["newNumberVector"](state, payload) {
      applyUpdate(state, payload);
    },
    ["newBLOBVector"](state, payload) {
      applyUpdate(state, payload);
    },
    ["delProperty"](state, payload) {
      if (payload.name) {
        const existing = state.devices[payload.device] ? JSON.parse(JSON.stringify(state.devices[payload.device])) : {};
        delete existing[payload.name];
        Vue.set(state.devices, payload.device, existing);
      } else {
        delete state.devices[payload.device];
      }
    },
    ["statusUpdate"](state, payload) {
      applyUpdate(state, payload);
    },
    ["indi_client_connected"](state) {
      state.status = "connected";
      state.error = null;

      this.dispatch("indi/subscribe");
    },
    ["indi_client_closed"](state) {
      state.status = "disconnected";
    },
    ["indi_client_error"](state, payload) {
      state.error = payload;
    },
    ["indi_client_reconnecting"](state) {
      state.status = "reconnecting";
    },
  },
  actions: {
    subscribe: ({dispatch}) => {
      dispatch("sendCommand", {type: "getProperties", data: {}});
      dispatch("sendCommand", {type: "enableBLOB", data: new enableBLOB("", "", "Never")});
    },

    sendCommand: (context, {type, data}) => {
      if (type.startsWith("new")) {
        context.commit("statusUpdate", {
          device: data.device,
          name: data.name,
          state: "Busy",
        });
      }

      Vue.prototype.$socket.send(JSON.stringify({
        type: type,
        payload: data,
      }));
    }
  },
};


server/index.js
const http = require('http');
const WebSocket = require('ws');
const static = require('node-static');
const enableGracefulShutdown = require('server-graceful-shutdown');

const { INDIClient, mapping, getProperties, enableBLOB } = require("indi-client");

const debug = process.env.DEBUG === "1";
const indiHost = process.env.INDI_HOST || "localhost";
const indiPort = process.env.INDI_PORT ? parseInt(process.env.INDI_PORT) : 7624;

const indiClient = new INDIClient(indiHost, indiPort);

const server = http.createServer();

const interfaceServer = new WebSocket.Server({ noServer: true });

var signals = {
  'SIGHUP': 1,
  'SIGINT': 2,
  'SIGTERM': 15
};

// Do any necessary shutdown logic for our application here
const shutdown = (signal, value) => {
  console.log("shutdown!");

  interfaceServer.close();
  indiClient.close();

  server.shutdown(() => {
    console.log(`server stopped by ${signal} with value ${value}`);
    process.exit(128 + value);
  });
};

// Create a listener for each of the signals that we want to handle
Object.keys(signals).forEach((signal) => {
  process.on(signal, () => {
    console.log(`process received a ${signal} signal`);
    shutdown(signal, signals[signal]);
  });
});

const sendJSON = (msg) => {
  if (debug) {
    console.log('sending message to interface', msg);
  }

  interfaceServer.clients.forEach(c => {
    c.send(JSON.stringify(msg));
  });
};

interfaceServer.on("connection", (ws) => {
  if (indiClient.isConnected()) {
    sendJSON({
      type: "indi_client_connected",
    });
  }

  ws.on("message", (msg) => {
    // Every message we get from the client should be forwarded to INDI.
    const parsed = JSON.parse(msg);

    let cls = mapping[parsed.type];
    if (!cls) {
      switch (parsed.type) {
        case "getProperties":
          cls = getProperties;
          break;
        case "enableBLOB":
          cls = enableBLOB;
          break;
      }
    }

    if (debug) {
      console.log(parsed);
    }

    const obj = cls.fromJSON(parsed.payload);

    indiClient.send(obj);
  });
});

Object.keys(mapping).forEach(k => {
  indiClient.on(k, (msg) => {
    sendJSON({
      type: k,
      payload: msg,
    });
  });
});

indiClient.on("connect", () => {
  console.log("connected to indi server");
  sendJSON({
    type: "indi_client_connected",
  });
});

indiClient.on("close", () => {
  console.log("indiclient closed");
  sendJSON({
    type: "indi_client_closed",
  });
});

indiClient.on("error", (err) => {
  sendJSON({
    type: "indi_client_error",
    payload: err,
  });
});

indiClient.on("reconnect", () => {
  console.log("indiclient reconnecting");
  sendJSON({
    type: "indi_client_reconnecting",
  });
});

var file = new static.Server("./static");

server.addListener("request", (req, res) => {
  file.serve(req, res);
});

server.on("upgrade", (req, socket, head) => {
  console.log('upgrade started', req.url);

  interfaceServer.handleUpgrade(req, socket, head, (ws) => {
    interfaceServer.emit("connection", ws, req);
  });
});

indiClient.connect();

enableGracefulShutdown(server, 1000);

server.listen(8081);
